### PR TITLE
HTTP API: validate payload size

### PR DESCRIPTION
**Pull request by dashevchenko**
## Pull Request description

Added spring filter to filter large rpc requests

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: common/transport/transport-api/src/main/java/org/thingsboard/server/common/transport/service/DefaultTransportService.java [hunk 1]
```diff
@@ -824,6 +827,21 @@ public void log(TransportProtos.SessionInfoProto sessionInfo, String msg) {
         process(sessionInfo, postTelemetryMsg, TransportServiceCallback.EMPTY);
     }
 
+    private boolean checkMsgSize(TransportProtos.SessionInfoProto sessionInfo, TransportServiceCallback<?> callback, CharSequence... payload) {
```
Comment (smatvienko-tb) [inline common/transport/transport-api/src/main/java/org/thingsboard/server/common/transport/service/DefaultTransportService.java:830]: The algorithm complexity should be O(1). Any higher complexity will not help prevent DDoS attacks (CPU overload).
You may sacrifice the precise calculation and use an estimate. e.g. charSequence length is enough. the real size may vary from 1 to 2 bytes per char. but it might be well enough protection.

But? please, read another comments before you made any changes.

File: common/transport/transport-api/src/main/java/org/thingsboard/server/common/transport/service/DefaultTransportService.java [hunk 2]
```diff
@@ -685,7 +688,7 @@ private void processTimeout(String requestId) {
 
     @Override
     public void process(TransportProtos.SessionInfoProto sessionInfo, TransportProtos.ToServerRpcRequestMsg msg, TransportServiceCallback<Void> callback) {
-        if (checkLimits(sessionInfo, msg, callback)) {
+        if (checkLimits(sessionInfo, msg, callback) && checkMsgSize(sessionInfo, callback, msg.getMethodName(), msg.getParams())) {
```
Comment (smatvienko-tb) [inline common/transport/transport-api/src/main/java/org/thingsboard/server/common/transport/service/DefaultTransportService.java:691]: ### Long story short
It is too late to verify the message size at this stage. In a DDoS case, you may already be OOM at this moment.
### TL;DR
You should check RAW payload size as early as possible.
The proto you check is already prepared be some code earlier.
The main goal is to not use any memory before the max_message_size  chack passed.

Transports uses some protocols and libraries. I believe some of them already allows to set a hard limit for payload in bytes on the lowest level possible. For example , max buffer size or so. 
Imagine you are receiving a 1GiG RPC request. The channel would rather not receive the whole 1GiG and then do some checks. You simply might not have the memory or network bandwidth to handle even a few parallel requests.

If the channel gives up immediately after exceeding the payload limit - it will be the best solution possible.

If the channel does not provide an option to check the msg limit - check the payload raw size as soon as possible before any transformations.

image

File: common/message/src/main/java/org/thingsboard/server/common/msg/tools/MaxMessageSizeExceededException.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import lombok.Getter;
+import org.thingsboard.server.common.data.id.DeviceId;
+
+public class MaxMessageSizeExceededException extends RuntimeException {
+
+    @Getter
+    private final DeviceId deviceId;
+
+    public MaxMessageSizeExceededException(DeviceId deviceId) {
+        super(String.format("Max message size exceeded for device [%s]", deviceId));
```
Comment (smatvienko-tb) [inline common/message/src/main/java/org/thingsboard/server/common/msg/tools/MaxMessageSizeExceededException.java:27]: I'm not a believer that we need `String.format` here. Just check how complex the implementation is, and you might want to use `string + string + string`

File: application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java [hunk 4]
```diff
@@ -74,6 +79,26 @@ public void testGetAttributes() throws Exception {
...
+    @Test
+    public void testReplyToCommandWithLargeResponse() throws Exception {
+        String errorResponse = doPostAsync("/api/v1/" + deviceCredentials.getCredentialsId() + "/rpc/5",
+                JacksonUtil.toString(createRpcResponsePayload(101)),
+                String.class,
+                status().is5xxServerError());
+        assertThat(errorResponse).isEqualTo("Max message size exceeded for device [" + device.getId() + "]");
+
+        doPostAsync("/api/v1/" + deviceCredentials.getCredentialsId() + "/rpc/5",
+                JacksonUtil.toString(createRpcResponsePayload(10)),
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java:91]: I would like to see the corner case. like Maximum allowed payload ok. And failed `max + 1` or any expected overflow

File: application/src/main/resources/thingsboard.yml [hunk 5]
```diff
@@ -951,6 +951,7 @@ transport:
     max_wrong_credentials_per_ip: "${TB_TRANSPORT_MAX_WRONG_CREDENTIALS_PER_IP:10}"
     # Timeout (in milliseconds) to expire block IP addresses
     ip_block_timeout: "${TB_TRANSPORT_IP_BLOCK_TIMEOUT:60000}"
+  max_message_size: "${TB_TRANSPORT_MAX_MESSAGE_SIZE:500000}" # max transport message size
```
Comment (smatvienko-tb) [inline application/src/main/resources/thingsboard.yml:954]: How did you come up with a 500k limit ?

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
Good start. But please, read my notes carefully and try to avoid any overhead

Please, also consider that some protocols already have a protection like MQTT

```
max_payload_size: "${NETTY_MAX_PAYLOAD_SIZE:65536}"
```

File: common/message/src/main/java/org/thingsboard/server/common/msg/tools/TbMaxPayloadSizeExceededException.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.common.msg.tools;
+
+public class TbMaxPayloadSizeExceededException extends RuntimeException {
+
+    public TbMaxPayloadSizeExceededException() {
+        super("Payload size exceeds the configured maximum");
```
Comment (smatvienko-tb) [inline common/message/src/main/java/org/thingsboard/server/common/msg/tools/TbMaxPayloadSizeExceededException.java:21]: Payload size exceeds the limit

File: application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java [hunk 7]
```diff
@@ -74,6 +79,54 @@ public void testGetAttributes() throws Exception {
...
+    }
+
+    private ObjectNode createRpcRequestPayload(int size) {
+        ObjectNode rpcResponseBody = JacksonUtil.newObjectNode();
+        char[] chars = new char[size-38];
+        Arrays.fill(chars, 'a');
+        rpcResponseBody.put("method", "get");
+        ObjectNode params = JacksonUtil.newObjectNode();
+        params.put("value", new String(chars));
+        rpcResponseBody.set("params", params);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java:125]: kind of overengineered.
Lets compose some simple string that are easy to read and understand like
```
{"method":"get","params":{"value":"aaaaaaaaaaaaaa"}}
```
the only "aaaaaaaa" can be generated. other just a static string . much clear as for me

File: application/src/main/java/org/thingsboard/server/exception/ThingsboardErrorResponseHandler.java [hunk 8]
```diff
@@ -184,6 +187,13 @@ private void handleRateLimitException(HttpServletResponse response, TbRateLimits
                         ThingsboardErrorCode.TOO_MANY_REQUESTS, HttpStatus.TOO_MANY_REQUESTS));
     }
 
+    private void handleMaxPayloadSizeExceededException(HttpServletResponse response, TbMaxPayloadSizeExceededException exception) throws IOException {
+        response.setStatus(HttpStatus.BAD_REQUEST.value());
```
Comment (smatvienko-tb) [inline application/src/main/java/org/thingsboard/server/exception/ThingsboardErrorResponseHandler.java:191]: Http has a special status for large payload
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
Now I believe it  works. 
Please, workout some minor stuff in my comment:

File: application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java [hunk 9]
```diff
@@ -15,31 +15,31 @@
  */
 package org.thingsboard.server.config;
 
+import com.amazonaws.HttpMethod;
```
Comment (smatvienko-tb) [inline application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java:18]: why Amazon is here?

File: application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java [hunk 10]
```diff
@@ -15,31 +15,31 @@
...
+import java.util.Set;
 
 @Slf4j
 @Component
 @RequiredArgsConstructor
 public class DeviceRpcRequestSizeFilter extends OncePerRequestFilter {
 
-    private final RequestMatcher uriMatcher = new AntPathRequestMatcher("/api/v1/*/rpc/**", HttpMethod.POST.name());
-    private final RequestMatcher matcher = new NegatedRequestMatcher(uriMatcher);
+    private final Set<String> urls = new HashSet<>(List.of("/api/v1/*/rpc/**", "/api/plugins/rpc/**", "/api/rpc/**"));
```
Comment (smatvienko-tb) [inline application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java:41]: immutable `List.of` is enough. Or use Set.of(). But you never search/modify by URL pattern. So the set makes no sense here

**Review by smatvienko-tb — APPROVED** [review]
look good to me

File: application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+import org.thingsboard.server.common.msg.tools.TbMaxPayloadSizeExceededException;
+import org.thingsboard.server.exception.ThingsboardErrorResponseHandler;
+
+import java.io.IOException;
+import java.util.List;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class DeviceRpcRequestSizeFilter extends OncePerRequestFilter {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java:37]: It's in application module, so this won't work in a microservice setup.

File: application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+    private final List<String> urls = List.of("/api/v1/*/rpc/**", "/api/plugins/rpc/**", "/api/rpc/**");
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java:39]: Why only these endpoints? Device can as well supply a very large telemetry, attributes, etc.

File: application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java (hunk 11 above)
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/config/DeviceRpcRequestSizeFilter.java:37]: After you cover more endpoints unrelated to RPC, this should be renamed to `RequestSizeFilter`.

File: application/src/main/resources/thingsboard.yml [hunk 13]
```diff
@@ -959,6 +959,8 @@ transport:
     request_timeout: "${HTTP_REQUEST_TIMEOUT:60000}"
     # HTTP maximum request processing timeout in milliseconds
     max_request_timeout: "${HTTP_MAX_REQUEST_TIMEOUT:300000}"
+    # RPC maximum request size
+    rpc_max_payload_size: "${HTTP_RPC_MAX_PAYLOAD_SIZE:65536}" # max rpc payload size in bytes
```
[the remark this answers is not in the document]
Comment (ViacheslavKlimov) [reply application/src/main/resources/thingsboard.yml:963]: Rename to `max_payload_size`

File: common/message/src/main/java/org/thingsboard/server/common/msg/tools/TbMaxPayloadSizeExceededException.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.common.msg.tools;
+
+public class TbMaxPayloadSizeExceededException extends RuntimeException {
```
Comment (ViacheslavKlimov) [inline common/message/src/main/java/org/thingsboard/server/common/msg/tools/TbMaxPayloadSizeExceededException.java:18]: "Tb" prefix is unnecessary.

**Review by ViacheslavKlimov — CHANGES_REQUESTED** [review]

File: common/transport/http/src/main/java/org/thingsboard/server/transport/http/DeviceApiController.java [hunk 15]
```diff
@@ -265,14 +276,22 @@ public DeferredResult<ResponseEntity> subscribeToCommands(
...
             @Parameter(description = ACCESS_TOKEN_PARAM_DESCRIPTION, required = true , schema = @Schema(defaultValue = "YOUR_DEVICE_ACCESS_TOKEN"))
             @PathVariable("deviceToken") String deviceToken,
             @Parameter(description = "RPC request id from the incoming RPC request", required = true , schema = @Schema(defaultValue = "123"))
             @PathVariable("requestId") Integer requestId,
             @io.swagger.v3.oas.annotations.parameters.RequestBody(description = "Reply to the RPC request, JSON. For example: {\"status\":\"success\"}", required = true)
-            @RequestBody String json) {
+            @RequestBody String json, HttpServletRequest httpServletRequest) {
+        if (httpServletRequest.getContentLength() > rpcMaxPayloadSize) {
+            throw new MaxPayloadSizeExceededException();
+        }
```
Comment (ViacheslavKlimov) [inline common/transport/http/src/main/java/org/thingsboard/server/transport/http/DeviceApiController.java:294]: Extract to method `checkPayloadSize` or `checkRequestSize`
